_posts/2020-05-01-native-command-mode-tools.adoc

```diff
@@ -0,0 +1,137 @@
+An full example of a working native-image build under windows can be found link:https://github.com/38leinaD/quarkus-native-app-starter/blob/master/.github/workflows/release-build-native-win64.yml[here].
+
+After triggering a release from the Github UI, the jobs for all target environments are triggered. The artifacts for all platforms are will be available within minutes and can be consumed by your users. Building, hosting: 100% free and no external service or registration needed. 
+
+image:github-actions.jpg[]
+
+== Command-mode for UI-based applications
+
+Another exciting thing about command-mode is that it not only allows to write nice command-line tools (using e.g. link:https://picocli.info/[Picocli], link:http://aeshell.github.io/[Aesh] or link:https://commons.apache.org/proper/commons-cli/[Apache Commons CLI]) but in general allows to have more control over the lifecycle of an application.
+With the possibiliy to embeed your HTML/Javascript frontend, we can even build link:https://www.electronjs.org/[Electron-style] desktop application with rich user interfaces for our tools. As even JSF most likely will be link:https://github.com/adminfaces/quarkus-admin-starter[coming to Quarkus], there is a great variety of choices how to build a frontend. 
+
+A very simple scaffold does not require much. If you have `quarkus-resteasy` as a dependency, the embeeded server will be started on port `8080` just like in any regular Quarkus backend application.
+As your tools might be running on a developer system alongside other app-servers or Quarkus instances, it is a good idea to choose a different port to not run into conflicts.
+
+Configuring `quarkus.http.port=0` in your `application.properties` will make Quarkus choose a random but free port.
+How to know from within your application which port it is? When injecting the config-property via `@ConfigProperty(name = "quarkus.http.port")`, you will get the assigned port.
+
+Assuming you have your front-end under `src/main/resources/META-INF/resources` (in my example it is plain HTML with Javascript) and using a very old Java API, you can spin up the default browser to your app like this: 
+
+[source, java]
+----
+@QuarkusMain
+public class HelloMain implements QuarkusApplication {
+
+    @ConfigProperty(name = "quarkus.http.port")
+    Integer assignedPort; <1>
+    
+    public static void main(String[] args) {
+        Quarkus.run(HelloMain.class, args);
+    }
+
+    @Override
+    public int run(String... args) throws Exception {
+        URI webappUri = new URI("http://localhost:" + assignedPort + "/index.html");
+        
+        Desktop.getDesktop().browse(webappUri);
```

Comment:
might want to add the key imports  on top of this..

Reply:
Done